File: src/algebra1_factoring_trinomials_easy.py
"""
algebra1 - factoring_trinomials (easy)
Generated: 2026-02-22T04:00:42.070028
"""

from problem_utils import *

def generate():
    problem_type = randint(1, 3)

    if problem_type == 1:
        # Both roots positive (x+r1)(x+r2) — varied range
        r1 = randint(1, 10)
        r2 = randint(1, 10)
        factored = (x + r1) * (x + r2)
        expr = expand(factored)
        ans = factor(expr)
        b_coeff = r1 + r2
        c_coeff = r1 * r2

        return problem(
            question=f"Factor the trinomial: ${latex(expr)}$",
            answer=ans,
            difficulty=(1000, 1100),
            topic="algebra1/factoring_trinomials",
            grading_mode="factor",
            solution=steps(
                f"Find two numbers that multiply to ${c_coeff}$ and add to ${b_coeff}$",
                f"The numbers are ${r1}$ and ${r2}$: ${r1} \\cdot {r2} = {c_coeff}$, ${r1}+{r2} = {b_coeff}$",
                f"${latex(ans)}$"
            )
        )

    elif problem_type == 2:
        # One positive, one negative root — wide range
        r1 = randint(1, 9)
        r2 = randint(1, 9)
        sign = choice([1, -1])
        factored = (x + sign*r1) * (x - sign*r2) if randint(0,1) else (x - r1) * (x + r2)
        expr = expand(factored)
        ans = factor(expr)
        b_coeff = int(expr.coeff(x, 1))
        c_coeff = int(expr.coeff(x, 0))

        return problem(
            question=f"Factor the trinomial: ${latex(expr)}$",
            answer=ans,
            difficulty=(1100, 1200),
            topic="algebra1/factoring_trinomials",
            grading_mode="factor",
            solution=steps(
                f"Find two numbers that multiply to ${c_coeff}$ and add to ${b_coeff}$",
                f"Since the constant is {'negative, the two numbers have opposite signs' if c_coeff < 0 else 'positive, the two numbers have the same sign'}",
                f"${latex(ans)}$"
            )
        )

    else:
        # Both roots negative (x-r1)(x-r2) — wide range
        r1 = randint(1, 10)
        r2 = randint(1, 10)
        factored = (x - r1) * (x - r2)
        expr = expand(factored)
        ans = factor(expr)
        b_coeff = -(r1 + r2)
        c_coeff = r1 * r2

        return problem(
            question=f"Factor the trinomial: ${latex(expr)}$",
            answer=ans,
            difficulty=(1200, 1300),
            topic="algebra1/factoring_trinomials",
            grading_mode="factor",
            solution=steps(
                f"Find two numbers that multiply to ${c_coeff}$ and add to ${b_coeff}$",
                f"Both numbers are negative: $(-{r1}) \\cdot (-{r2}) = {c_coeff}$, $(-{r1})+(-{r2}) = {b_coeff}$",
                f"${latex(ans)}$"
            )
        )

emit(generate())

File: src/algebra1_factoring_trinomials_easy.jl
# algebra1 - factoring_trinomials (easy)
# Generated: 2026-03-08T20:00:12.720352

include(joinpath(@__DIR__, "..", "julia", "src", "ProblemUtils.jl"))
using .ProblemUtils

@script x begin
    set_topic!("algebra1/factoring_trinomials")
    
    # Choose problem type randomly
    prob_type = choice(1:4)
    
    if prob_type == 1
        # Simple factorable trinomial: (x + r1)(x + r2)
        r1 = nonzero(-12, 12)
        r2 = nonzero(-12, 12)
        expr = expand((x + r1) * (x + r2))
        
        b_coeff = r1 + r2
        c_coeff = r1 * r2
        
        problem(
            question="Factor the trinomial: \$$(tex(expr))\$",
            answer=expr,
            difficulty=(700, 900),
            grading_mode="factor",
            solution=steps(
                sol("Given", expr),
                "Find two numbers that multiply to $(c_coeff) and add to $(b_coeff)",
                "The numbers are $(r1) and $(r2)",
                sol("Factored form", (x + r1) * (x + r2))
            ),
            time=90
        )
        
    elseif prob_type == 2
        # Trinomial with leading coefficient: a(x + r1)(x + r2)
        a = nonzero(-5, 5)
        r1 = nonzero(-10, 10)
        r2 = nonzero(-10, 10)
        expr = expand(a * (x + r1) * (x + r2))
        
        problem(
            question="Factor completely: \$$(tex(expr))\$",
            answer=expr,
            difficulty=(900, 1100),
            grading_mode="factor",
            solution=steps(
                sol("Given", expr),
                "Factor out common factor if any, then find binomial factors",
                "Look for factors of the form \$($(a)x + p)(x + q)\$ or similar",
                sol("Factored form", a * (x + r1) * (x + r2))
            ),
            time=120
        )
        
    elseif prob_type == 3
        # Perfect square trinomial: (x + r)^2
        r = nonzero(-15, 15)
        expr = expand((x + r)^2)
        
        problem(
            question="Factor the perfect square trinomial: \$$(tex(expr))\$",
            answer=expr,
            difficulty=(800, 1000),
            grading_mode="factor",
            solution=steps(
                sol("Given", expr),
                "Recognize this is a perfect square: \$a^2 + 2ab + b^2 = (a + b)^2\$",
                "Here \$a = x\$ and \$b = $(r)\$",
                sol("Factored form", (x + r)^2)
            ),
            time=90
        )
        
    else
        # Difference of squares disguised or simple factorable
        r1 = nonzero(-12, 12)
        r2 = -r1
        a = nonzero(1, 4)
        expr = expand(a * (x + r1) * (x + r2))
        
        problem(
            question="Factor completely: \$$(tex(expr))\$",
            answer=expr,
            difficulty=(850, 1050),
            grading_mode="factor",
            solution=steps(
                sol("Given", expr),
                "Notice the middle term coefficient is 0 (or factor out \$$(a)\$ first)",
                "This is a difference of squares: \$a^2 - b^2 = (a-b)(a+b)\$",
                sol("Factored form", a * (x + r1) * (x + r2))
            ),
            time=100
        )
    end
end